api/PayrollProcessor.Functions.Api/Features/Departments/DepartmentEmployeeTrigger.cs

```diff
@@ -60,37 +62,48 @@ public class DepartmentEmployeeTrigger
                     return;
 
                 default:
+                    log.LogWarning("Queue message with {eventName} cannot be handled", eventName);
+
                     return;
             }
         }
 
-        private Task HandleEmployeeCreation(CloudQueueMessage queueMessage, ILogger log)
-        {
-            var message = QueueMessageHandler.FromQueueMessage<EmployeeCreation>(queueMessage);
-
-            return queryDispatcher.Dispatch(new EmployeeQuery(message.EmployeeId))
-                .Bind(employee => commandDispatcher.Dispatch(new DepartmentEmployeeCreateCommand(employee, idGenerator.Generate())))
-                .Bind(departmentEmployee => apiClient.SendNotification(nameof(EmployeeUpdatesQueue), departmentEmployee))
-                .Match(
-                    _ => log.LogInformation(""),
-                    () => log.LogError(""),
-                    e => log.LogError(e, ""));
-        }
+        private Task HandleEmployeeCreation(CloudQueueMessage queueMessage, ILogger log) =>
+            queueMessage
+                .Apply(QueueMessageHandler.FromQueueMessage<EmployeeCreation>)
+                .Apply(message => queryDispatcher.Dispatch(new EmployeeQuery(message.EmployeeId))
+                    .DoIfNoneOrFail(
+                        () => log.LogError("Could not find Employee {employeeId}", message.EmployeeId),
+                        ex => log.LogError(ex, "Could not query for Employee {employeeId}", message.EmployeeId))
+                    .Bind(
+                        employee => commandDispatcher.Dispatch(new DepartmentEmployeeCreateCommand(employee, idGenerator.Generate()))
+                            .DoIfFail(ex => log.LogError(ex, "Could not create Department Employee for {@employee}", employee))
```

Comment:
We nest the `DoIfNoneOrFail()` and `DoIfFail()` inside the `.Bind()` call so that these logging calls will only be executed for failures of the operations executed on success of the previous operations.

Example: If the above query fails, we don't want to log on line 76-77 _and_ on line 80, but if the query succeeds and the command fails, we want to log on line 80.

Placing the `.DoIfFail()` as a chained call of `.Bind()` on line 78 would result in any failure of a previous operation to get logged here as well.